File: docs/scripts/fig_architecture.py
"""Figure 3 for the PAB design document: storage & reporting architecture.

Draws the flow of PAB outputs from the analysis layer through storage to the
community-facing reporting channels:

    BING fits  ->  SQLite DB (extracted values)        ->  Reporting
               ->  Nautilus/NSF S3 (chains + figures)  ->  Reporting + downloads
               ->  Zenodo snapshot (DOI, pab_version)

Usage:
    python docs/scripts/fig_architecture.py [out.png]

Writes docs/figures/fig_architecture.png by default.
"""
import os
import sys
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.patches import FancyBboxPatch, FancyArrowPatch

DEFAULT_OUT = os.path.join(os.path.dirname(__file__), "..", "figures",
                           "fig_architecture.png")

# palette
GREY = "#ededed"
BLUE = "#d9ebff"
GREEN = "#d9f7dc"
ORANGE = "#ffedcc"
EDGE = "#333333"


def _box(ax, x, y, w, h, label, face):
    """Draw a centered, rounded box (x,y = lower-left, data coords 0..100)."""
    ax.add_patch(FancyBboxPatch(
        (x, y), w, h, boxstyle="round,pad=0.6,rounding_size=2.0",
        linewidth=1.3, edgecolor=EDGE, facecolor=face, zorder=2))
    ax.text(x + w / 2, y + h / 2, label, ha="center", va="center",
            fontsize=10.5, zorder=3)
    return (x, y, w, h)


def _arrow(ax, xy_from, xy_to):
    ax.add_patch(FancyArrowPatch(
        xy_from, xy_to, arrowstyle="-|>", mutation_scale=14,
        linewidth=1.3, color="#555555", zorder=1,
        shrinkA=2, shrinkB=2))


def main(out_path=DEFAULT_OUT):
    fig, ax = plt.subplots(figsize=(9.0, 4.4))
    ax.set_xlim(0, 100)
    ax.set_ylim(0, 100)
    ax.axis("off")

    # Source
    src = _box(ax, 3, 58, 22, 18, "BING fits\n(per matchup)", GREY)

    # Storage tier
    db = _box(ax, 37, 74, 27, 16, "SQLite DB\nextracted values", BLUE)
    s3 = _box(ax, 37, 50, 27, 16, "Nautilus / NSF S3\nchains + figures", GREEN)
    zen = _box(ax, 37, 26, 27, 16, "Zenodo snapshot\nDOI · pab_version", ORANGE)

    # Reporting tier
    rep = _box(ax, 74, 50, 23, 26,
               "Reporting\nreadthedocs\n+ Bokeh\n+ downloads", GREY)

    # arrows: source -> storage
    _arrow(ax, (src[0] + src[2], 69), (db[0], 83))
    _arrow(ax, (src[0] + src[2], 65), (s3[0], 58))
    # storage -> Zenodo (DB and S3 feed the citable snapshot)
    _arrow(ax, (db[0] + 6, db[1]), (zen[0] + 8, zen[1] + zen[3]))
    _arrow(ax, (s3[0] + 6, s3[1]), (zen[0] + 16, zen[1] + zen[3]))
    # storage -> reporting
    _arrow(ax, (db[0] + db[2], 82), (rep[0], 70))
    _arrow(ax, (s3[0] + s3[2], 58), (rep[0], 60))

    ax.set_title("PAB storage & reporting architecture", fontsize=13)

    fig.tight_layout()
    out_path = os.path.abspath(out_path)
    os.makedirs(os.path.dirname(out_path), exist_ok=True)
    fig.savefig(out_path, dpi=110)
    print(f"Wrote {out_path} ({os.path.getsize(out_path)/1024:.0f} KB)")


if __name__ == "__main__":
    out = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_OUT
    main(out)

File: docs/scripts/fig_architecture.m
% fig_architecture.m
% Figure 3 for the PAB design document: the storage / reporting architecture.
%
% Draws the flow of PAB outputs from the analysis layer through storage to the
% community-facing reporting channels:
%
%   BING fits ->  SQLite DB (extracted values)         ->  Reporting (readthedocs + Bokeh)
%             ->  Nautilus/NSF S3 (chains + figures)   ->  Reporting + downloads
%             ->  Zenodo snapshot (DOI, pab_version)
%
% Usage (MATLAB):
%   >> run('docs/scripts/fig_architecture.m')
% Writes docs/figures/fig_architecture.png
%
% NOTE: This script was authored on a workstation without MATLAB installed, so
% it has NOT been executed/verified here. Run it in MATLAB to produce the PNG.

function fig_architecture()
    fig = figure('Color', 'w', 'Position', [100 100 900 460]);
    ax = axes('Position', [0 0 1 1]); axis(ax, [0 100 0 100]); axis off; hold on;

    % --- helper for a labelled box ---
    function h = box(x, y, w, h_, label, face)
        rectangle('Position', [x y w h_], 'Curvature', 0.12, ...
            'FaceColor', face, 'EdgeColor', [0.2 0.2 0.2], 'LineWidth', 1.2);
        text(x + w/2, y + h_/2, label, 'HorizontalAlignment', 'center', ...
            'VerticalAlignment', 'middle', 'FontSize', 11, 'Interpreter', 'none');
        h = [x y w h_];
    end

    % --- helper for an arrow between box edges ---
    function arrow(x1, y1, x2, y2)
        annotation('arrow', 'Units', 'normalized', ...
            'Position', normpos(x1, y1, x2 - x1, y2 - y1), ...
            'HeadStyle', 'vback2', 'LineWidth', 1.2, 'Color', [0.3 0.3 0.3]);
    end
    function p = normpos(x, y, dx, dy)
        % data (0..100) -> normalized figure coords for annotation()
        p = [x/100, y/100, dx/100, dy/100];
    end

    blue   = [0.85 0.92 1.00];
    green  = [0.85 0.97 0.86];
    orange = [1.00 0.93 0.80];
    grey   = [0.92 0.92 0.92];

    % Source
    box( 4, 60, 22, 16, sprintf('BING fits\n(per matchup)'), grey);

    % Storage tier
    box(38, 74, 26, 16, sprintf('SQLite DB\nextracted values'), blue);
    box(38, 52, 26, 16, sprintf('Nautilus/NSF S3\nchains + figures'), green);
    box(38, 30, 26, 16, sprintf('Zenodo snapshot\nDOI, pab_version'), orange);

    % Reporting tier
    box(74, 52, 22, 24, sprintf('Reporting\nreadthedocs\n+ Bokeh\n+ downloads'), grey);

    % Arrows: source -> storage
    arrow(26, 70, 38, 82);   % fits -> SQLite
    arrow(26, 66, 38, 60);   % fits -> S3
    arrow(51, 52, 51, 46);   % SQLite/S3 -> Zenodo (downward)

    % storage -> reporting
    arrow(64, 82, 74, 70);   % SQLite -> reporting
    arrow(64, 60, 74, 62);   % S3 -> reporting

    title('PAB storage & reporting architecture', 'FontSize', 13);

    outdir = fullfile(fileparts(mfilename('fullpath')), '..', 'figures');
    if ~exist(outdir, 'dir'); mkdir(outdir); end
    out = fullfile(outdir, 'fig_architecture.png');
    exportgraphics(fig, out, 'Resolution', 110);
    fprintf('Wrote %s\n', out);
end
